File: ui/aria_utils.js
/**
 * ARIA Utilities
 * 
 * This module provides utilities for adding ARIA attributes to components
 * to improve accessibility.
 */

/**
 * Enhances icon-only buttons with proper ARIA labels
 * 
 * @param {Object} component - The button component to enhance
 * @param {string} label - The accessible label
 * @param {Object} options - Additional options
 * @returns {Object} The enhanced component
 */
function addAriaLabelToIconButton(component, label, options = {}) {
    // Default options
    const defaultOptions = {
        role: 'button',
        includeTooltip: true,
        tooltipPlacement: 'top'
    };
    
    const mergedOptions = { ...defaultOptions, ...options };
    
    // Check if the component already has props
    if (!component.props) {
        component.props = {};
    }
    
    // Add ARIA attributes to component props
    component.props['aria-label'] = label;
    
    if (mergedOptions.role) {
        component.props.role = mergedOptions.role;
    }
    
    // Add tooltip if requested and if dbc is available
    if (mergedOptions.includeTooltip && typeof window !== 'undefined') {
        try {
            const dbc = window.dash_bootstrap_components;
            
            if (component.props.id && dbc) {
                // Create unique tooltip ID
                const tooltipId = `${component.props.id}-tooltip`;
                
                // Create tooltip
                const tooltip = {
                    type: 'Tooltip',
                    namespace: 'dash_bootstrap_components',
                    props: {
                        target: component.props.id,
                        placement: mergedOptions.tooltipPlacement,
                        children: label
                    }
                };
                
                // Return both the button and tooltip
                return [component, tooltip];
            }
        } catch (e) {
            console.warn('Could not add tooltip to button', e);
        }
    }
    
    return component;
}

/**
 * Enhances a checkbox with proper ARIA attributes
 * 
 * @param {Object} component - The checkbox component to enhance
 * @param {string} label - The accessible label text
 * @returns {Object} The enhanced component
 */
function enhanceCheckbox(component, label) {
    if (!component.props) {
        component.props = {};
    }
    
    component.props['aria-label'] = label;
    component.props.role = 'checkbox';
    
    if (component.props.checked !== undefined) {
        component.props['aria-checked'] = !!component.props.checked;
    }
    
    return component;
}

/**
 * Creates visually hidden text for screen readers
 * 
 * @param {string} text - The text content for screen readers
 * @returns {Object} A component that's visually hidden but accessible to screen readers
 */
function createScreenReaderOnly(text) {
    return {
        type: 'span',
        namespace: 'dash_html_components',
        props: {
            children: text,
            style: {
                position: 'absolute',
                width: '1px',
                height: '1px',
                padding: '0',
                margin: '-1px',
                overflow: 'hidden',
                clip: 'rect(0, 0, 0, 0)',
                whiteSpace: 'nowrap',
                borderWidth: '0'
            }
        }
    };
}

/**
 * Add appropriate ARIA roles to tables
 * 
 * @param {Object} tableComponent - The table component to enhance
 * @param {Object} options - Additional options
 * @returns {Object} The enhanced table component
 */
function enhanceDataTable(tableComponent, options = {}) {
    // Set role="table" on the table element
    tableComponent.props.role = 'table';
    
    // Add caption if provided
    if (options.caption) {
        // Find or create a caption element
        let caption;
        const children = Array.isArray(tableComponent.props.children) 
            ? tableComponent.props.children 
            : [tableComponent.props.children];
            
        const captionIndex = children.findIndex(c => 
            c && c.type && c.type === 'caption');
            
        if (captionIndex >= 0) {
            // Update existing caption
            caption = children[captionIndex];
            caption.props.children = options.caption;
        } else {
            // Create new caption
            caption = {
                type: 'caption',
                namespace: 'dash_html_components',
                props: {
                    children: options.caption
                }
            };
            
            // Add caption to the children
            children.unshift(caption);
            tableComponent.props.children = children;
        }
    }
    
    // Recursively process table children to add appropriate roles
    function processTableChildren(element) {
        if (!element || typeof element !== 'object') return element;
        
        // Add appropriate role based on element type
        if (element.type === 'thead') {
            element.props.role = 'rowgroup';
        } else if (element.type === 'tbody') {
            element.props.role = 'rowgroup';
        } else if (element.type === 'tr') {
            element.props.role = 'row';
        } else if (element.type === 'th') {
            element.props.role = 'columnheader';
        } else if (element.type === 'td') {
            element.props.role = 'cell';
        }
        
        // Process children recursively
        if (element.props && element.props.children) {
            if (Array.isArray(element.props.children)) {
                element.props.children = element.props.children.map(processTableChildren);
            } else {
                element.props.children = processTableChildren(element.props.children);
            }
        }
        
        return element;
    }
    
    // Process children
    if (tableComponent.props.children) {
        if (Array.isArray(tableComponent.props.children)) {
            tableComponent.props.children = tableComponent.props.children.map(processTableChildren);
        } else {
            tableComponent.props.children = processTableChildren(tableComponent.props.children);
        }
    }
    
    return tableComponent;
}

// Export the utilities
export {
    addAriaLabelToIconButton,
    enhanceCheckbox,
    createScreenReaderOnly,
    enhanceDataTable
};

File: ui/aria_utils.py
"""
ARIA Utilities

This module provides utilities for adding ARIA attributes to components
to improve accessibility.
"""

from dash import html
import dash_bootstrap_components as dbc


def add_aria_label_to_icon_button(component, label, options=None):
    """
    Enhances icon-only buttons with proper ARIA labels.

    Args:
        component: The button component to enhance
        label: The accessible label
        options: Additional options (role, includeTooltip, tooltipPlacement)

    Returns:
        The enhanced component or a list containing the component and tooltip
    """
    options = options or {}

    # Default options
    default_options = {
        "role": "button",
        "include_tooltip": True,
        "tooltip_placement": "top",
    }

    # Merge options
    for key, value in default_options.items():
        if key not in options:
            options[key] = value

    # Add ARIA attributes
    if not hasattr(component, "className"):
        component.className = ""

    component.className += " has-aria-label"
    component.role = options["role"]

    # Convert component.children to a list if it's not already
    if not isinstance(component.children, list):
        component.children = [component.children]

    # Add aria-label attribute using the correct Dash property format
    if hasattr(component, "children"):
        # For HTML components, aria-label should be passed as a constructor argument
        # Since we can't modify it after creation, we'll use title instead
        pass

    # Use title attribute for accessibility (works for screen readers)
    if not hasattr(component, "title") or component.title is None:
        component.title = label

    # Add tooltip if requested
    if options["include_tooltip"] and hasattr(component, "id"):
        tooltip = dbc.Tooltip(
            label, target=component.id, placement=options["tooltip_placement"]
        )
        return [component, tooltip]

    return component


def enhance_checkbox(component, label):
    """
    Enhances a checkbox with proper ARIA attributes.

    Args:
        component: The checkbox component to enhance
        label: The accessible label text

    Returns:
        The enhanced component
    """
    if not hasattr(component, "className"):
        component.className = ""

    component.className += " has-aria-label"
    component.role = "checkbox"

    if hasattr(component, "checked"):
        component["aria-checked"] = bool(component.checked)

    return component


def create_screen_reader_only(text):
    """
    Creates visually hidden text for screen readers.

    Args:
        text: The text content for screen readers

    Returns:
        A component that's visually hidden but accessible to screen readers
    """
    return html.Span(
        text,
        style={
            "position": "absolute",
            "width": "1px",
            "height": "1px",
            "padding": "0",
            "margin": "-1px",
            "overflow": "hidden",
            "clip": "rect(0, 0, 0, 0)",
            "whiteSpace": "nowrap",
            "borderWidth": "0",
        },
    )


def enhance_data_table(table_component, options=None):
    """
    Add appropriate ARIA roles to tables.

    Args:
        table_component: The table component to enhance
        options: Additional options like caption

    Returns:
        The enhanced table component
    """
    options = options or {}

    # Set role="table" on the table element
    table_component.role = "table"

    # Add caption if provided
    if "caption" in options:
        # Find or create a caption element
        caption = None
        children = (
            table_component.children
            if isinstance(table_component.children, list)
            else [table_component.children]
        )

        for i, child in enumerate(children):
            if isinstance(child, html.Caption):
                caption = child
                caption.children = options["caption"]
                break

        if not caption:
            caption = html.Caption(options["caption"])
            children.insert(0, caption)
            table_component.children = children

    # Recursively process table children to add appropriate roles
    def process_table_children(element):
        if not element or not isinstance(element, html.Base):
            return element

        # Add appropriate role based on element type
        if isinstance(element, html.Thead):
            element.role = "rowgroup"  # type: ignore[attr-defined]
        elif isinstance(element, html.Tbody):
            element.role = "rowgroup"  # type: ignore[attr-defined]
        elif isinstance(element, html.Tr):
            element.role = "row"  # type: ignore[attr-defined]
        elif isinstance(element, html.Th):
            element.role = "columnheader"  # type: ignore[attr-defined]
        elif isinstance(element, html.Td):
            element.role = "cell"  # type: ignore[attr-defined]

        # Process children recursively
        if hasattr(element, "children"):
            if isinstance(element.children, list):
                element.children = [
                    process_table_children(child) for child in element.children
                ]
            else:
                element.children = process_table_children(element.children)

        return element

    # Process children
    if hasattr(table_component, "children"):
        if isinstance(table_component.children, list):
            table_component.children = [
                process_table_children(child) for child in table_component.children
            ]
        else:
            table_component.children = process_table_children(table_component.children)

    return table_component
